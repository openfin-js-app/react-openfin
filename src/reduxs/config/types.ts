import { SvgIconProps } from '@material-ui/core/SvgIcon';
import { TextFieldProps } from '@material-ui/core/TextField';
import {TimePickerModalProps} from "material-ui-pickers/TimePicker/TimePickerModal";
import {DatePickerModalProps} from "material-ui-pickers/DatePicker/DatePickerModal";

export enum MuiTheme{
    LIGHT='light',
    DARK='dark',
}

export enum I18Language{
    en_US='en-US',
    zh_CN ='zh-CN',
}

export enum FieldType{
    CUSTOM_FIELD = 1,
    STRING = 2,
    DATETIME = 3,
    DATE = 4,
    TIME = 5,
    NUMBER = 6,
    CURRENCY = 7,
    TITLE = 8,
    SUBHEADING = 9,
    BODY1 = 10,
}

export interface IConfigField {
    _type:FieldType;
    _label:string;
    _name?:string;
    _defaultValue?:any;
    _cols?:number;
    _rows?:number;
    _props?:TextFieldProps|TimePickerModalProps|DatePickerModalProps;
    _custom?:any;
    [key:string]:any;
    [key:number]:any;
}

export interface IConfigTab {
    _order?:number;
    _label:string;
    _name:string;
    _svgUrl:string;
    _svgUrl_dark?:string;
    _icon?:React.ComponentType<SvgIconProps>;
    _fields:IConfigField[];
    _fieldLabels?:string;
    [key:string]:any;
    [key:number]:any;
}

export interface IConfigDexie {
    id?:number,
    tabName:string,
    fieldName:string,
    value:any,
    version?:number,
}

export interface IConfigResetOption {
    tabName?:string
}

export interface IConfigUpdateOneFieldOption {
    name:string,
    value:any,
}

export interface IConfigDoUpdateOneFieldOption {
    tabName:string,
    fieldName:string,
    value:any,
}

export interface IConfigUpdateGlobalFilterStrOption {
    configGlobalFilterString:string,
}

export interface IConfigRuntimeState{
    application?:Partial<{
        theme:MuiTheme,
        language:I18Language,
        newWinTop:number,
        newWinLeft:number,
        newWinWidth:number,
        newWinHeight:number,
        newWindDeltaLeft:number,
        newWindDeltaHeight:number,
    }>,
    sample?:Partial<{
        strValue:string,
        crrValue:number,
        dateValue:Date,
        timeValue:Date,
        datetimeValue:Date,
    }>
}

export interface IConfigState extends Partial<IConfigRuntimeState>{
    configGlobalFilterString:string;
    _tabs:IConfigTab[];
    [key:string]:any;
    [key:number]:any;
}