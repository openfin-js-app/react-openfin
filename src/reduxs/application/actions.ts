import { createAction, ActionFunctionAny, Action } from 'redux-actions';
import {WindowOptions} from 'redux-openfin';

import { IApplicationNewSnackbarOption, IApplicationCloseSnackBarOption } from './types';
import * as React from "react";

export const APPLICATION_SET_LOADING_MSG:string = 'APPLICATION_SET_LOADING_MSG';
export const APPLICATION_STARTED:string = 'APPLICATION_STARTED';
export const APPLICATION_CHILD_STARTED:string = 'APPLICATION_CHILD_STARTED';
export const APPLICATION_NOTIFICATION_STARTED:string = 'APPLICATION_NOTIFICATION_STARTED';
export const APPLICATION_READY:string = 'APPLICATION_READY';
export const APPLICATION_DRAWER_TOGGLE:string = 'APPLICATION_DRAWER_TOGGLE';
export const APPLICATION_TOGGLE_WINDOW_STATE:string = 'APPLICATION_TOGGLE_WINDOW_STATE';

export const applicationSetLoadingMsg:ActionFunctionAny<Action<{loadingMsg:string}>> =
    createAction<{loadingMsg:string},string>(APPLICATION_SET_LOADING_MSG,(loadingMsg:string)=>({loadingMsg}));
export const applicationStarted:ActionFunctionAny<Action<void>> = createAction(APPLICATION_STARTED);
export const applicationChildStarted:ActionFunctionAny<Action<void>> = createAction(APPLICATION_CHILD_STARTED);
export const applicationNotificationStarted:ActionFunctionAny<Action<void>> = createAction(APPLICATION_NOTIFICATION_STARTED);
export const applicationReady:ActionFunctionAny<Action<void>> = createAction(APPLICATION_READY);
export const applicationDrawerToggle:ActionFunctionAny<Action<void>> = createAction(APPLICATION_DRAWER_TOGGLE);
export const applicationToogleWindowState:ActionFunctionAny<Action<void>> = createAction(APPLICATION_TOGGLE_WINDOW_STATE);

// snackbar

export const APPLICATION_NEW_SNACKBAR='APPLICATION_NEW_SNACKBAR';
export const APPLICATION_SET_SNACKBAR_STATUS='APPLICATION_SET_SNACKBAR_STATUS';
export const APPLICATION_PROCESS_SNACKBAR_QUEUE='APPLICATION_PROCESS_SNACKBAR_QUEUE';
export const APPLICATION_CLOSE_SNACKBAR='APPLICATION_CLOSE_SNACKBAR';

export const applicationNewSnackbar:ActionFunctionAny<Action<IApplicationNewSnackbarOption>>
    = createAction(APPLICATION_NEW_SNACKBAR, (option:IApplicationNewSnackbarOption)=>(option));
export const applicationSetSnackbarStatus = createAction<{snackBarOpen:boolean},boolean>(APPLICATION_SET_SNACKBAR_STATUS, (snackBarOpen:boolean)=>({snackBarOpen}));
export const applicationProcessSnackbarQueue:ActionFunctionAny<Action<{}>>
    = createAction(APPLICATION_PROCESS_SNACKBAR_QUEUE);
export const applicationCloseSnackbar:ActionFunctionAny<Action<IApplicationCloseSnackBarOption>>
    = createAction(APPLICATION_CLOSE_SNACKBAR, (event: React.SyntheticEvent<any>, reason: string)=>({event,reason}));


// snap dock
export const APPLICATION_UPDATE_DOCK_STATUS = 'APPLICATION_UPDATE_DOCK_STATUS';
export const applicationUpdateDockStatus:ActionFunctionAny<Action<{docked:boolean}>>
    = createAction(APPLICATION_UPDATE_DOCK_STATUS, (docked:boolean)=>({docked}));

// launch bar
export const APPLICATION_LAUNCH_BAR_TOGGLE = 'APPLICATION_LAUNCH_BAR_TOGGLE';
export const APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE = 'APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE';
export const APPLICATION_LAUNCH_NEW_WINDOW = 'APPLICATION_LAUNCH_NEW_WINDOW';
export const APPLICATION_LAUNCH_BAR_CLOSE = 'APPLICATION_LAUNCH_BAR_CLOSE';
export const applicationLaunchBarToggle:ActionFunctionAny<Action<void>> = createAction(APPLICATION_LAUNCH_BAR_TOGGLE);
export const applicationLaunchBarToggleCollapse:ActionFunctionAny<Action<void>> = createAction(APPLICATION_LAUNCH_BAR_TOGGLE_COLLAPSE);
export const applicationLaunchNewWindow:ActionFunctionAny<Action<Partial<WindowOptions>>>
    = createAction(APPLICATION_LAUNCH_NEW_WINDOW,(appJson:Partial<WindowOptions>)=>(appJson));
export const applicationLaunchBarClose:ActionFunctionAny<Action<void>> = createAction(APPLICATION_LAUNCH_BAR_CLOSE);

// network
export const APPLICATION_NETWORK_ONLINE = 'APPLICATION_NETWORK_ONLINE';
export const applicationNetworkOnline:ActionFunctionAny<Action<void>> = createAction(APPLICATION_NETWORK_ONLINE);
export const APPLICATION_NETWORK_OFFLINE = 'APPLICATION_NETWORK_OFF_LINE';
export const applicationNetworkOffline:ActionFunctionAny<Action<void>> = createAction(APPLICATION_NETWORK_OFFLINE);
