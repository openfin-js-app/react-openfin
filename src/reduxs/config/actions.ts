import {createAction, ActionFunctionAny, Action} from 'redux-actions';

import {
    IConfigDexie,
    IConfigResetOption,
    IConfigUpdateOneFieldOption,
    IConfigUpdateGlobalFilterStrOption,
    IConfigDoUpdateOneFieldOption
} from './types';

export const CONFIG_RESET:string = 'CONFIG_RESET';

// redux
export const CONFIG_DO_UPDATE_ONE_FIELD:string = 'CONFIG_DO_UPDATE_ONE_FIELD';
export const CONFIG_UPDATE_GLOBAL_FILTER_STR:string = 'CONFIG_UPDATE_GLOBAL_FILTER_STR';
export const configReset:ActionFunctionAny<Action<IConfigResetOption>>
    = createAction(CONFIG_RESET, (option)=>(option));
export const configDoUpdateOneField:ActionFunctionAny<Action<IConfigDoUpdateOneFieldOption>>
    = createAction(CONFIG_DO_UPDATE_ONE_FIELD, (option)=>(option));
export const configUpdateGlobalFilterStr:ActionFunctionAny<Action<IConfigUpdateGlobalFilterStrOption>>
    = createAction(CONFIG_UPDATE_GLOBAL_FILTER_STR, (option)=>(option));

// saga
export const CONFIG_LOAD_FROM_DEXIE:string = 'CONFIG_LOAD_FROM_DEXIE';
export const configLoadFromDexie:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_LOAD_FROM_DEXIE,);
export const CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE:string = 'CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE';
export const configDoUpdateOneFieldInDexie:ActionFunctionAny<Action<IConfigDexie>>
    = createAction(CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE, (option)=>(option));

export const CONFIG_UPDATE_ONE_FIELD:string = 'CONFIG_UPDATE_ONE_FIELD';
export const configUpdateOneField:ActionFunctionAny<Action<IConfigUpdateOneFieldOption>>
    = createAction(CONFIG_UPDATE_ONE_FIELD, (option)=>(option));



// optional actions depending on the config application values

export const CONFIG_UPDATE_NEW_WINDOW_POSITION:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION';
export const CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA';
export const CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP';
export const CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT:string = 'CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT';
export const configUpdateNewWindowPosition:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION);
export const configUpdateNewWindowPositionAddDelta:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA);
export const configUpdateNewWindowPositionResetTop:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP);
export const configUpdateNewWindowPositionResetLeft:ActionFunctionAny<Action<void>>
    = createAction(CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT);