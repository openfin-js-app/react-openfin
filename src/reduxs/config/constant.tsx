import * as React from 'react';
import InputAdornment from '@material-ui/core/InputAdornment';

import initState from '../../init';

import { IConfigTab, IConfigField, MuiTheme, FieldType } from './types';

const configTabs:IConfigTab[]=[
    {
        _order: 0 ,
        _label: 'application.label',
        _name:'application',
        _svgUrl: void 0,
        _fields:[
            {
                _type:FieldType.SUBHEADING,
                _label:'application.theme.label',
                _cols: 2,
                _rows:1,
            },
            {
                _type:FieldType.CUSTOM_FIELD,
                _label:'Theme',
                _custom:<React.Fragment/>,
                _name:'theme',
                _defaultValue:MuiTheme.DARK,
                _cols:10,
                _rows:1,
            },
            {
                _type:FieldType.SUBHEADING,
                _label:'application.language.label',
                _cols: 2,
                _rows:1,
            },
            {
                _type:FieldType.CUSTOM_FIELD,
                _label:'language',
                _custom:<React.Fragment/>,
                _name:'language',
                _defaultValue:'en-US',
                _cols:10,
                _rows:1,
            },
            {
                _type:FieldType.SUBHEADING,
                _label:'application.newWinPos.label',
                _cols: 12,
            },
            {
                _type:FieldType.NUMBER,
                _label:'application.newWinPos.newWinTop',
                _name:'newWinTop',
                _defaultValue:initState.config.newWindowTop,
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
            {
                _type:FieldType.NUMBER,
                _label:'application.newWinPos.newWinLeft',
                _name:'newWinLeft',
                _defaultValue:initState.config.newWindowLeft,
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
            {
                _type:FieldType.NUMBER,
                _label:'application.newWinPos.newWinWidth',
                _name:'newWinWidth',
                _defaultValue: initState.config.newWindowWidth,
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
            {
                _type:FieldType.NUMBER,
                _label:'application.newWinPos.newWinHeight',
                _name:'newWinHeight',
                _defaultValue:initState.config.newWindowHeight,
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
            {
                _type:FieldType.NUMBER,
                _label:'application.newWinPos.newWindDeltaLeft',
                _name:'newWindDeltaLeft',
                _defaultValue:initState.config.newWindowDeltaLeft,
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
            {
                _type:FieldType.NUMBER,
                _label:'application.newWinPos.newWindDeltaHeight',
                _name:'newWindDeltaHeight',
                _defaultValue:initState.config.newWindowDeltaHeight,
                _props:{
                    InputProps:{
                        endAdornment:<InputAdornment position={'end'}>Pixel</InputAdornment>
                    }
                }
            },
        ]
    },
    {
        _order: 1 ,
        _label: 'about.label',
        _name:'about',
        _svgUrl: void 0,
        _fields:[
            {
                _type:FieldType.CUSTOM_FIELD,
                _label:'About openfin starter',
                _custom:<React.Fragment/>,
                _cols:12,
                _rows:12,
            }
        ]
    }
];

export default configTabs;
