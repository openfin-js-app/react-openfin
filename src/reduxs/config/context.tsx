import * as React from 'react';

import { Omit } from '../../utils/typeHelper';

import { I18Language, IConfigRuntimeState } from './types';

interface IWithConfig {
    config:Partial<IConfigRuntimeState>,
    actions:Partial<{
        onToggleThemeField:()=>void,
        onUpdateLangField:(lang:I18Language)=>void,
    }>
}

interface IWithConfigContext {
    configContext?:Partial<IWithConfig>
}

export const ConfigContext = React.createContext<Partial<IWithConfig>|null>(null);

const { Provider, Consumer } = ConfigContext;

export type WithConfigContext = IWithConfigContext;

export const ConfigContextProvider = Provider;
export const ConfigContextConsumer = Consumer;

export const withConfigContext:<
    P extends WithConfigContext
>(Component: React.ComponentType<P>)=>React.FunctionComponent<Omit<P,WithConfigContext>> =

    <P extends WithConfigContext>(Component: React.ComponentType<P>)=> {
        return function ComponentWithConfig(props:Omit<P,WithConfigContext>){
            return (
                <ConfigContextConsumer>
                    {value =>
                        // @ts-ignore
                        (<Component {...props} configContext={value}/>)}
                </ConfigContextConsumer>
            );}
};