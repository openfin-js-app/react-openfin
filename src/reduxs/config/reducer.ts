import { handleActions, Action } from 'redux-actions';

import initState from '../../init'

import {
    CONFIG_RESET,CONFIG_DO_UPDATE_ONE_FIELD,
    CONFIG_UPDATE_GLOBAL_FILTER_STR,
    CONFIG_EXTEND_CUST_STATE,

    CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA,
    CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP,
    CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT,
} from './actions';
import {
    IConfigTab, IConfigField, FieldType,
    IConfigResetOption, IConfigDoUpdateOneFieldOption, IConfigState
} from './types';

import configTabs  from './constant';

export function buildDefaultConfigState(configTabs: IConfigTab[]):IConfigState{
    const result:IConfigState = {
        configGlobalFilterString:'',
        _tabs:configTabs,
    };
    configTabs.forEach(oneTab =>{
        oneTab._fieldLabels='';
        result[oneTab._name]={};
        oneTab._fields.forEach(oneField =>{
            oneTab._fieldLabels = oneTab._fieldLabels+oneField._label;
            if(oneField._name && (oneField._defaultValue || oneField._defaultValue==="")){
                result[oneTab._name][oneField._name]=oneField._defaultValue;
            }
        })
    });
    return result;
}

export const defaultState:IConfigState = buildDefaultConfigState(configTabs);

export const buildInitState = (parentWindowState?:IConfigState) =>{
    return defaultState;
}

const reducerMap:{[key:string]: (state:IConfigState, action?:Action<any>)=>IConfigState} = {
    [CONFIG_RESET]:(state,action)=>{
        const {tabName} = action.payload as IConfigResetOption;
        if(tabName){
            return {
                ...state,
                tabName:defaultState[tabName],
            }
        }else{
            return defaultState;
        }
    },
    [CONFIG_DO_UPDATE_ONE_FIELD]:(state,action)=>{
        const {tabName, fieldName, value} = action.payload as IConfigDoUpdateOneFieldOption;
        return {
            ...state,
            [tabName]:{
                ...state[tabName],
                [fieldName]:value,
            }
        };
    },
    [CONFIG_UPDATE_GLOBAL_FILTER_STR]:(state,action)=>{
        const { configGlobalFilterString } = action.payload;
        return{
            ...state,
            configGlobalFilterString,
        };
    },
    [CONFIG_EXTEND_CUST_STATE]: (state,action)=>{
        const custState = action.payload;

        if (custState.application){
            delete custState.application;
        }

        return {
            ...state,
            ...custState,
        }

    },
    [CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA]:(state,action)=>{
        const applicationConfig = state.application;

        return {
            ...state,
            application:{
                ...state.application,
                newWinTop: applicationConfig.newWinTop+ applicationConfig.newWindDeltaHeight,
                newWinLeft: applicationConfig.newWinLeft+ applicationConfig.newWindDeltaLeft,
            }
        };
    },
    [CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP]:(state,action)=>({
        ...state,
        application:{
            ...state.application,
            newWinTop: initState.config.newWindowTop,
        }
    }),
    [CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT]:(state,action)=>({
        ...state,
        application:{
            ...state.application,
            newWinLeft: initState.config.newWindowLeft,
        }
    }),
}

const reducer = (state:IConfigState, action:Action<any>):IConfigState => {

    if (action.type && reducerMap[action.type]){
        return reducerMap[action.type](state,action);
    }else{
        return state;
    }

}

export default reducer;