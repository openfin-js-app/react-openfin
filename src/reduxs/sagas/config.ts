import { call, delay, select, put, putResolve, take, takeLatest, takeEvery } from 'redux-saga/effects';
import { System } from "redux-openfin";

import {
    CONFIG_LOAD_FROM_DEXIE,
    CONFIG_UPDATE_NEW_WINDOW_POSITION,
    CONFIG_UPDATE_ONE_FIELD,
    CONFIG_SELECT_ONE_FIELD,
    CONFIG_REMOVE_ONE_FIELD,
    CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE,

    configDoUpdateOneField,
    configDoUpdateOneFieldInDexie,
    configSelectOneFieldRes,
    configUpdateNewWindowPositionAddDelta,
    configUpdateNewWindowPositionResetLeft,
    configUpdateNewWindowPositionResetTop,

    IConfigDexie,
    IConfigUpdateOneFieldOption,
    IConfigSelectOneFieldOption,
    IConfigSelectOneFieldResPayload,
    IConfigRemoveOneFieldOption,
} from '..';

import {
    findAllOfCurrentVersion,saveOrUpdateOneByTabNameFieldName, removeOneByTabNameAndFieldName,
} from '../../dexies/configDao';

export const getNewWindowTop = state => state.config.application.newWinTop;
export const getNewWindowLeft = state => state.config.application.newWinLeft;
export const getNewWindowWidth = state => state.config.application.newWinWidth;
export const getNewWindowHeight = state => state.config.application.newWinHeight;

export function* handleConfigLoadFromDexie() {
    const configs : IConfigDexie[] = yield call(findAllOfCurrentVersion);
    for (const config of configs){
        yield put(configDoUpdateOneField({
            tabName:config.tabName,
            fieldName:config.fieldName,
            value:config.value,
        }))
    }
}

export function* handleConfigUpdateOneField(action) {
    const {name,value} = action.payload as IConfigUpdateOneFieldOption;
    const paths = name.split('.');
    if (paths.length === 2){
        yield putResolve(configDoUpdateOneField({
            tabName:paths[0],
            fieldName:paths[1],
            value,
        }));
        yield put(configDoUpdateOneFieldInDexie({
            tabName:paths[0],
            fieldName:paths[1],
            value,
        }))
    }
}

export function* handleConfigSelectOneField(action){
    const { tabName, fieldName } = action.payload as IConfigSelectOneFieldOption;

    const value = yield select( state => {
        if ( tabName in state && fieldName in state[tabName]){
            return state[tabName][fieldName];
        }else{
            return null;
        }
    });

    const result:IConfigSelectOneFieldResPayload={
        tabName,fieldName,value,
        userObj: 'userObj' in action.payload ? action.payload.userObj : null,
    };

    putResolve(configSelectOneFieldRes(result));
}

export function* handleConfigRemoveOneField(action){
    const { tabName, fieldName } = action.payload as IConfigRemoveOneFieldOption;
    yield call(removeOneByTabNameAndFieldName,tabName, fieldName);

    if ('value' in action.payload){
        yield putResolve(configDoUpdateOneField({
            tabName,
            fieldName,
            value:action.payload.value,
        }));
    }else{
        yield putResolve(configDoUpdateOneField({
            tabName,
            fieldName,
            value:{},
        }));
    }

}

export function* handleConfigUpdateOneFieldInDexie(action) {
    const {tabName, fieldName, value} = action.payload as IConfigDexie;
    yield call(saveOrUpdateOneByTabNameFieldName,tabName,fieldName,value);
    yield delay(800);
}

export function* handleConfigUpdateNewWindowPosition() {
    const newWinWidth = yield select(getNewWindowWidth);
    const newWinHeight = yield select(getNewWindowHeight);
    const newWinTop = yield select(getNewWindowTop);
    const newWinLeft = yield select(getNewWindowLeft);

    const monitorInfoAction = yield call(System.asyncs.getMonitorInfo,System.actions.getMonitorInfo({}));
    const virtualScreen = monitorInfoAction.payload.virtualScreen;

    // console.log("configUpdateNewWindowPosition",monitorInfoAction,virtualScreen);

    if (
        ((newWinLeft+newWinWidth)<virtualScreen.right) &&
        ((newWinTop+newWinHeight)<virtualScreen.bottom)
    ){
        yield putResolve(configUpdateNewWindowPositionAddDelta());
    }else{
        if ((newWinLeft+newWinWidth)>=virtualScreen.right){
            yield putResolve(configUpdateNewWindowPositionResetLeft());
        }
        if ((newWinTop+newWinHeight)>=virtualScreen.bottom){
            yield putResolve(configUpdateNewWindowPositionResetTop());
        }
    }

}

export default function* () {
    yield takeEvery(CONFIG_LOAD_FROM_DEXIE, handleConfigLoadFromDexie);
    yield takeEvery(CONFIG_UPDATE_ONE_FIELD, handleConfigUpdateOneField);
    yield takeEvery(CONFIG_SELECT_ONE_FIELD, handleConfigSelectOneField);
    yield takeEvery(CONFIG_REMOVE_ONE_FIELD, handleConfigRemoveOneField);
    yield takeLatest(CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE, handleConfigUpdateOneFieldInDexie);
    yield takeLatest(CONFIG_UPDATE_NEW_WINDOW_POSITION,handleConfigUpdateNewWindowPosition);
}