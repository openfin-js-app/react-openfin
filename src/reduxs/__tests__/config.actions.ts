import * as Actions from '../config/actions';

describe('Config actions',()=>{

    it('configReset action',()=>{
        expect(Actions.configReset({})).toMatchSnapshot();
    })

    it('configDoUpdateOneField action',()=>{
        expect(Actions.configDoUpdateOneField({
            tabName:'tabName',
            fieldName:'fieldName',
            value:'value',
        })).toMatchSnapshot();
    })

    it('configUpdateGlobalFilterStr action',()=>{
        expect(Actions.configUpdateGlobalFilterStr('string')).toMatchSnapshot();
    })

    it('configLoadFromDexie action',()=>{
        expect(Actions.configLoadFromDexie()).toMatchSnapshot();
    })

    it('configDoUpdateOneFieldInDexie action',()=>{
        expect(Actions.configDoUpdateOneFieldInDexie({
            id:1,
            tabName:'tableName',
            fieldName:'fieldName',
            value:'value',
            version:1
        })).toMatchSnapshot();
    })

    it('configUpdateOneField action',()=>{
        expect(Actions.configUpdateOneField('str1','str2',{})).toMatchSnapshot();
    })

    it('configSelectOneField action',()=>{
        expect(Actions.configSelectOneField({tabName:'tabName', fieldName:'fieldName', userObj:{}})).toMatchSnapshot();
    })

    it('configRemoveOneField action',()=>{
        expect(Actions.configRemoveOneField({tabName:'tabName', fieldName:'fieldName', value:{}})).toMatchSnapshot();
    })

    it('configSelectOneFieldRes action',()=>{
        expect(Actions.configSelectOneFieldRes({tabName:'tabName', fieldName:'fieldName', userObj:{}, value:{}})).toMatchSnapshot();
    })

    it('configUpdateNewWindowPosition action',()=>{
        expect(Actions.configUpdateNewWindowPosition()).toMatchSnapshot();
    })

    it('configUpdateNewWindowPositionAddDelta action',()=>{
        expect(Actions.configUpdateNewWindowPositionAddDelta()).toMatchSnapshot();
    })

    it('configUpdateNewWindowPositionResetTop action',()=>{
        expect(Actions.configUpdateNewWindowPositionResetTop()).toMatchSnapshot();
    })

    it('configUpdateNewWindowPositionResetLeft action',()=>{
        expect(Actions.configUpdateNewWindowPositionResetLeft()).toMatchSnapshot();
    })

})