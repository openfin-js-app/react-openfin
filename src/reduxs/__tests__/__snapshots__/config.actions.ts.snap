// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config actions configDoUpdateOneField action 1`] = `
Object {
  "payload": Object {
    "fieldName": "fieldName",
    "tabName": "tabName",
    "value": "value",
  },
  "type": "REACT_OPENFIN::CONFIG_DO_UPDATE_ONE_FIELD",
}
`;

exports[`Config actions configDoUpdateOneFieldInDexie action 1`] = `
Object {
  "payload": Object {
    "fieldName": "fieldName",
    "id": 1,
    "tabName": "tableName",
    "value": "value",
    "version": 1,
  },
  "type": "REACT_OPENFIN::CONFIG_DO_UPDATE_ONE_FIELD_IN_DEXIE",
}
`;

exports[`Config actions configLoadFromDexie action 1`] = `
Object {
  "type": "REACT_OPENFIN::CONFIG_LOAD_FROM_DEXIE",
}
`;

exports[`Config actions configRemoveOneField action 1`] = `
Object {
  "payload": Object {
    "fieldName": "fieldName",
    "tabName": "tabName",
    "value": Object {},
  },
  "type": "REACT_OPENFIN_REQ::CONFIG_REMOVE_ONE_FIELD",
}
`;

exports[`Config actions configReset action 1`] = `
Object {
  "payload": Object {},
  "type": "REACT_OPENFIN::CONFIG_RESET",
}
`;

exports[`Config actions configSelectOneField action 1`] = `
Object {
  "payload": Object {
    "fieldName": "fieldName",
    "tabName": "tabName",
    "userObj": Object {},
  },
  "type": "REACT_OPENFIN_REQ::CONFIG_SELECT_ONE_FIELD",
}
`;

exports[`Config actions configSelectOneFieldRes action 1`] = `
Object {
  "payload": Object {
    "fieldName": "fieldName",
    "tabName": "tabName",
    "userObj": Object {},
    "value": Object {},
  },
  "type": "REACT_OPENFIN_RES::CONFIG_SELECT_ONE_FIELD_RES",
}
`;

exports[`Config actions configUpdateGlobalFilterStr action 1`] = `
Object {
  "payload": Object {
    "configGlobalFilterString": "string",
  },
  "type": "REACT_OPENFIN_REQ::CONFIG_UPDATE_GLOBAL_FILTER_STR",
}
`;

exports[`Config actions configUpdateNewWindowPosition action 1`] = `
Object {
  "type": "REACT_OPENFIN::CONFIG_UPDATE_NEW_WINDOW_POSITION",
}
`;

exports[`Config actions configUpdateNewWindowPositionAddDelta action 1`] = `
Object {
  "type": "REACT_OPENFIN::CONFIG_UPDATE_NEW_WINDOW_POSITION_ADD_DELTA",
}
`;

exports[`Config actions configUpdateNewWindowPositionResetLeft action 1`] = `
Object {
  "type": "REACT_OPENFIN::CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_LEFT",
}
`;

exports[`Config actions configUpdateNewWindowPositionResetTop action 1`] = `
Object {
  "type": "REACT_OPENFIN::CONFIG_UPDATE_NEW_WINDOW_POSITION_RESET_TOP",
}
`;

exports[`Config actions configUpdateOneField action 1`] = `
Object {
  "payload": Object {
    "name": "str1.str2",
    "value": Object {},
  },
  "type": "REACT_OPENFIN_REQ::CONFIG_UPDATE_ONE_FIELD",
}
`;
